workflow-runtime/src/commonMain/kotlin/com/squareup/workflow1/internal/RealRenderContext.kt

```diff
@@ -70,6 +84,96 @@ internal class RealRenderContext<out PropsT, StateT, OutputT>(
     sideEffectRunner.runningSideEffect(key, sideEffect)
   }
 
+  override fun <ResultT : Any> remember(
+    key: String,
+    resultType: KClass<ResultT>,
+    vararg inputs: Any?,
+    calculation: () -> ResultT
+  ): ResultT {
+    checkNotFrozen()
+    return rememberStore.remember(key, resultType, inputs = inputs, calculation)
+  }
+
+  override fun eventHandler(
+    name: String,
+    update: WorkflowAction<
+      @UnsafeVariance PropsT,
+      StateT,
+      @UnsafeVariance OutputT
+      >.Updater.() -> Unit
+  ): () -> Unit {
+    val handler = { actionSink.send(action("eventHandler:$name", update)) }
+    return if (stableEventHandlers) {
+      class Box0 {
+        lateinit var delegate: () -> Unit
+        fun fire() = delegate()
+      }
+
+      val box = remember("stableEventHandler:$name") { Box0() }
+      box.delegate = handler
+      box::fire
+    } else {
+      handler
+    }
+  }
+
+  override fun <EventT : Any> eventHandler(
+    name: String,
+    eventType: KClass<EventT>,
+    update: WorkflowAction<
+      @UnsafeVariance PropsT,
+      StateT,
+      @UnsafeVariance OutputT
+      >.Updater.(EventT) -> Unit
+  ): (EventT) -> Unit {
+    val handler: (EventT) -> Unit = { event ->
+      actionSink.send(action("eventHandler:$name") { update(event) })
+    }
+    return if (stableEventHandlers) {
+      class Box1<E> {
+        lateinit var delegate: (E) -> Unit
+        fun fire(e: E) = delegate(e)
+      }
+
+      val key = "stableEventHandler:$name+${eventType.simpleName}"
```

Comment:
- [ ] Find a way to use `expect` / `actual` so that we can use `qualifiedName` on non-JavaScript platforms.